### PR TITLE
Implement Islamic Civil Calendar

### PR DESCRIPTION
**Pull request by sotam1069**
Stems from #3673

### CODE REVIEW

File: components/calendar/src/astronomy.rs [hunk 1]
```diff
@@ -978,10 +977,15 @@ impl Astronomical {
 
     #[allow(dead_code, clippy::unwrap_used, clippy::eq_op)]
     pub(crate) fn moonlag(date: Moment, location: Location) -> Option<f64> {
```
Comment (Manishearth) [inline components/calendar/src/astronomy.rs:979]: nit: please document why this would return None (and also what this does in general)

Comment (sotam1069) [reply components/calendar/src/astronomy.rs:979]: fixed and added reference

File: components/calendar/src/islamic.rs [hunk 2]
```diff
@@ -308,224 +303,1142 @@ impl DateTime<IslamicObservational> {
...
+        next(RataDie::new(tau as i64), Self::adjusted_saudi_criterion)
+    }
+
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L6996
+    #[allow(clippy::unwrap_used)]
+    fn saudi_islamic_from_fixed(date: RataDie) -> Date<UmmalQura> {
+        let crescent = Self::saudi_new_month_on_or_before(date);
+        let elapsed_months =
+            libm::round((crescent - FIXED_ISLAMIC_EPOCH_FRIDAY) as f64 / MEAN_SYNODIC_MONTH);
+        let year = (1.0 + div_rem_euclid_f64(elapsed_months, 12.0).0) as i32;
```
Comment (sffc) [inline components/calendar/src/islamic.rs:773]: Nit: Fix the saturation

File: components/calendar/src/islamic.rs [hunk 3]
```diff
@@ -308,224 +303,1142 @@ impl DateTime<IslamicObservational> {
...
+impl Calendar for UmmalQura {
+    type DateInner = UmmalQuraDateInner;
+    fn date_from_codes(
+        &self,
+        era: types::Era,
+        year: i32,
+        month_code: types::MonthCode,
+        day: u8,
+    ) -> Result<Self::DateInner, CalendarError> {
+        let year = if era.0 == tinystr!(16, "ah") {
```
Comment (sffc) [inline components/calendar/src/islamic.rs:595]: Nit: Era code

File: components/calendar/src/islamic.rs [hunk 4]
```diff
@@ -308,224 +303,1142 @@ impl DateTime<IslamicObservational> {
...
+    fn is_leap_year(year: i32) -> bool {
+        div_rem_euclid(14 + 11 * year, 30).1 < 11
+    }
+
+    fn last_month_day_in_year(year: i32) -> (u8, u8) {
+        if Self::is_leap_year(year) {
+            (12, 30)
+        } else {
+            (12, 29)
+        }
```
[the remark this answers is not in the document]
Comment (sotam1069) [reply components/calendar/src/islamic.rs:338]: Yes, if the year is a leap year, the last month contains 30 days.

File: components/calendar/src/islamic.rs [hunk 5]
```diff
@@ -308,224 +303,1142 @@ impl DateTime<IslamicObservational> {
...
+
+    fn date_from_codes(
+        &self,
+        era: types::Era,
         year: i32,
-        month: u8,
+        month_code: types::MonthCode,
         day: u8,
+    ) -> Result<Self::DateInner, CalendarError> {
+        let year = if era.0 == tinystr!(16, "ah") {
```
[the remark this answers is not in the document]
Comment (sotam1069) [reply components/calendar/src/islamic.rs:352]: done.

File: components/calendar/src/islamic.rs (hunk 2 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs:773]: fixed!

File: components/calendar/src/islamic.rs (hunk 3 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs:595]: done.

File: components/calendar/src/astronomy.rs [hunk 6]
```diff
@@ -979,6 +979,8 @@ impl Astronomical {
     }
 
     #[allow(dead_code, clippy::unwrap_used, clippy::eq_op)]
+    // Time between sunset and moonset on date at location. Returns None if there is no sunset or moonset on said date.
```
Comment (echeran) [inline components/calendar/src/astronomy.rs:982]: nit: comments go above the annotations. change in this file & in islamic.rs

(that keeps the code together with other code)

File: components/calendar/src/islamic.rs [hunk 7]
```diff
@@ -582,6 +578,248 @@ impl UmmAlQura {
...
+        )
+    }
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2090
+    #[allow(clippy::unwrap_used)]
+    fn islamic_from_fixed(date: RataDie) -> Date<IslamicCivil> {
+        let year = div_rem_euclid_f64(
+            ((date.to_f64_date() - FIXED_ISLAMIC_EPOCH_FRIDAY.to_f64_date()) * 30.0) + 10646.0,
+            10631.0,
+        )
+        .0 as i32;
```
Comment (sffc) [inline components/calendar/src/islamic.rs:739]: Issue: saturate instead of `as i32`

File: components/calendar/src/islamic.rs [hunk 8]
```diff
@@ -582,6 +578,248 @@ impl UmmAlQura {
...
+        let prior_days = date.to_f64_date()
+            - ((Self::fixed_from_islamic(IslamicCivilDateInner(
+                ArithmeticDate::new_from_lunar_ordinals(year, 1, 1).unwrap(), // Safe unwrap due to hardcoded values,
+            )))
+            .to_f64_date());
+        let month = div_rem_euclid_f64((prior_days * 11.0) + 330.0, 325.0).0 as u8;
```
Comment (sffc) [inline components/calendar/src/islamic.rs:745]: Issue: Not immediately clear that this is within the bounds of a u8.  Please add an explanatory comment.

File: components/calendar/src/islamic.rs [hunk 9]
```diff
@@ -582,6 +578,248 @@ impl UmmAlQura {
...
+                + div_rem_euclid_f64(month as f64, 2.0).0 as i64
+                + day,
+        )
+    }
+    // Lisp code reference: https://github.com/EdReingold/calendar-code2/blob/main/calendar.l#L2090
+    #[allow(clippy::unwrap_used)]
+    fn islamic_from_fixed(date: RataDie) -> Date<IslamicCivil> {
+        let year = div_rem_euclid_f64(
+            ((date.to_f64_date() - FIXED_ISLAMIC_EPOCH_FRIDAY.to_f64_date()) * 30.0) + 10646.0,
+            10631.0,
```
Comment (sffc) [inline components/calendar/src/islamic.rs:737]: Nit: You can do this whole calculation in i64 without converting to float, and `date - FIXED_ISLAMIC_EPOCH_FRIDAY` should "just work"

File: components/calendar/src/islamic.rs [hunk 10]
```diff
@@ -582,6 +578,248 @@ impl UmmAlQura {
...
+                ArithmeticDate::new_from_lunar_ordinals(year, 1, 1).unwrap(), // Safe unwrap due to hardcoded values,
+            )))
+            .to_f64_date());
+        let month = div_rem_euclid_f64((prior_days * 11.0) + 330.0, 325.0).0 as u8;
+        let day = (date.to_f64_date()
+            - (Self::fixed_from_islamic(IslamicCivilDateInner(
+                ArithmeticDate::new_from_lunar_ordinals(year, month, 1).unwrap(), // Safe unwrap,
+            ))
+            .to_f64_date())
+            + 1.0) as u8;
```
Comment (sffc) [inline components/calendar/src/islamic.rs:751]: Issue: Another `as u8` that needs explanation

File: components/calendar/src/islamic.rs [hunk 11]
```diff
@@ -582,6 +578,248 @@ impl UmmAlQura {
...
+            _ => 0,
+        }
+    }
+
+    fn months_for_every_year(_year: i32) -> u8 {
+        12
+    }
+
+    fn days_in_provided_year(_year: i32) -> u32 {
+        355
```
Comment (echeran) [inline components/calendar/src/islamic.rs:601]: Leap years give an extra day in the year, so this should be a conditional based on whether it's a leap year

File: components/calendar/src/islamic.rs (hunk 7 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs]: fixed!

File: components/calendar/src/islamic.rs (hunk 9 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs]: true, I removed it thanks.

File: components/calendar/src/islamic.rs (hunk 8 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs]: Added a comment explaining why.

File: components/calendar/src/islamic.rs (hunk 10 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs]: done.

File: components/calendar/src/islamic.rs (hunk 11 above)
Comment (sotam1069) [reply components/calendar/src/islamic.rs]: done.

File: components/calendar/src/astronomy.rs (hunk 6 above)
Comment (sotam1069) [reply components/calendar/src/astronomy.rs:980]: Done in both files, thanks

**Review by sffc — APPROVED** [review]

**Pull request merged**